web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,288 @@
+        assert tracestatePropertiesMap != null;
+
+        StringBuffer outboundTracestate = new StringBuffer();
+        for (Map.Entry<String, String> entry : tracestatePropertiesMap.entrySet()) {
+            if (!entry.getKey().equals(AZURE_TRACEPARENT_COMPONENT_INITIAL)) {
+                outboundTracestate.append(entry);
+                outboundTracestate.append(",");
+            }
+        }
+        if (outboundTracestate.length() > 0) {
+            outboundTracestate.deleteCharAt(outboundTracestate.length()-1);
+        }
+
+        if (sourceAppId != null && sourceAppId.length() > 0) {
+            outboundTracestate.append(",");
+            outboundTracestate.append(AZURE_TRACEPARENT_COMPONENT_INITIAL).append("=").append(sourceAppId);
+        }
+        return new Tracestate(outboundTracestate.toString());
+    }
+
+
+    /**
+     * Extracts the appId/roleName out of Tracestate and compares it with the current appId. It then
+     * generates the appropriate source or target.
+     */
+    private static String generateSourceTargetCorrelation(String instrumentationKey, Map<String, String> tracestatePropertiesMap) {
+
+        assert instrumentationKey != null;
+        assert tracestatePropertiesMap != null;
+
+        String appId = tracestatePropertiesMap.get(AZURE_TRACEPARENT_COMPONENT_INITIAL);
+
+        String myAppId = InstrumentationKeyResolver.INSTANCE.resolveInstrumentationKey(instrumentationKey);
+
+        //it's possible the appId returned is null (e.g. async task is still pending or has failed). In this case, just
+        //return and let the next request resolve the ikey.
+        if (myAppId == null) {
+            InternalLogger.INSTANCE.trace("Could not generate source/target correlation as the appId could not be resolved (e.g. task may be pending or failed)");
+            return null;
+        }
+
+        // if the current appId and the incoming appId are send null
+        String result = null;
+        if (appId != null && !appId.equals(myAppId)) {
+            result = appId;
+        }
+
+        return result;
+    }
+
+    /**
+     * Parses the tracestate to generate a Map of vendor and their value property.
+     * @param tracestate
+     * @return
+     */
+    private static Map<String, String> getPropertiesMap(Tracestate tracestate) {
+
+        String tracestateAsString = tracestate.toString();
+        String[] vendorProperties = tracestateAsString.split(",");
+        Map<String, String> properties = new HashMap<>();
+
+        for (String s: vendorProperties) {
+            String[] keyval = s.split("=");
+            properties.put(keyval[0], keyval[1]);
+        }
+        return properties;
+    }
+
+    public static String retriveTracestate() {
+        //check if context is null - no correlation will happen
+        if (ThreadContext.getRequestTelemetryContext() == null) {
+            InternalLogger.INSTANCE.warn("No correlation wil happen, Thread context is null");
```

Comment:
typo: "wil" -> "will"

Also, shouldn't it read "RequestTelemetryContext" instead of "Thread context"?

Reply:
so the `ThreadContext` is the wrapper class which stores the RequestTelemetry in InherritableThreadLocal. When the threads change we want to retrive the context from ThreadLocal.

Reply:
I know, but the ThreadContext (call the static instance `T`) provides the RequestTelemetryContext (call the instance `R`)... `T.get -> R` and this checks if `R` is null but the messages says `T` is null.